src/material/sort/sort-header.ts

```diff
@@ -164,6 +164,10 @@ export class MatSortHeader extends _MatSortHeaderMixinBase
           }
 
           changeDetectorRef.markForCheck();
+
+          if ((<Sort>change) && (<Sort>change).active === this.id) {
```

Comment:
I've encountered a similar version of this bug that is not fixed by this change - if the inputs on MatSort change, it propagates to here via _stateChanges rather than sortChange.

From my own testing, I think you can fix both issues by instead doing:

if (_sort.active === this.id) {

Reply:
Thinking a bit further, probably what you'd want is something like:

parentSort = change || _sort;

if (parentSort && (<Sort>parentSort).active === this.id) {